// @deepstream/client/dist/src/record/write-ack-service.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var sinon_1 = require("sinon");
var mocks_1 = require("../test/mocks");
var constants_1 = require("../constants");
var write_ack_service_1 = require("./write-ack-service");
var utils_1 = require("../util/utils");
describe('Write Ack Notifier', function () {
    var topic = constants_1.TOPIC.RECORD;
    var action = constants_1.RECORD_ACTION.CREATEANDPATCH;
    var name = 'record';
    var services;
    var writeAckService;
    var callbackSpy;
    beforeEach(function () {
        services = mocks_1.getServicesMock();
        writeAckService = new write_ack_service_1.WriteAcknowledgementService(services);
        callbackSpy = sinon_1.spy();
    });
    afterEach(function () {
        services.verify();
    });
    it('cant\'t send request when client is offline', function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    services.connection.isConnected = false;
                    services.connectionMock
                        .expects('sendMessage')
                        .never();
                    writeAckService.send({
                        topic: topic,
                        action: action,
                        name: name
                    }, callbackSpy);
                    return [4 /*yield*/, utils_1.PromiseDelay(1)];
                case 1:
                    _a.sent();
                    sinon_1.assert.calledOnce(callbackSpy);
                    sinon_1.assert.calledWithExactly(callbackSpy, constants_1.EVENT.CLIENT_OFFLINE, name);
                    return [2 /*return*/];
            }
        });
    }); });
    it('calls callbacks with error message when connection is lost', function () { return __awaiter(void 0, void 0, void 0, function () {
        var messageBody;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    messageBody = {
                        topic: topic,
                        action: action,
                        name: name
                    };
                    writeAckService.send(messageBody, callbackSpy);
                    writeAckService.send(messageBody, callbackSpy);
                    services.simulateConnectionLost();
                    return [4 /*yield*/, utils_1.PromiseDelay(1)];
                case 1:
                    _a.sent();
                    sinon_1.assert.calledTwice(callbackSpy);
                    sinon_1.assert.calledWithExactly(callbackSpy, constants_1.EVENT.CLIENT_OFFLINE);
                    return [2 /*return*/];
            }
        });
    }); });
    it('sends correct messages with different correlationsId for each call', function () {
        var messageBody = {
            topic: topic,
            action: action,
            name: name,
        };
        services.connectionMock
            .expects('sendMessage')
            .once()
            .withExactArgs(Object.assign({}, messageBody, { action: action, correlationId: '1', isWriteAck: true }));
        services.connectionMock
            .expects('sendMessage')
            .once()
            .withExactArgs(Object.assign({}, messageBody, { action: action, correlationId: '2', isWriteAck: true }));
        writeAckService.send(messageBody, function () { });
        writeAckService.send(messageBody, function () { });
    });
    describe('receiving', function () {
        var correlationId = '1';
        var message;
        beforeEach(function () {
            message = {
                topic: topic,
                action: action,
                name: name
            };
            writeAckService.send(Object.assign({}, message), callbackSpy);
        });
        it('logs error for unknown acknowledgements', function () { return __awaiter(void 0, void 0, void 0, function () {
            var msg;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        msg = {
                            topic: topic,
                            action: action,
                            name: name,
                            correlationId: '123'
                        };
                        writeAckService.recieve(msg);
                        return [4 /*yield*/, utils_1.PromiseDelay(1)];
                    case 1:
                        _a.sent();
                        sinon_1.assert.notCalled(callbackSpy);
                        return [2 /*return*/];
                }
            });
        }); });
        it('calls ack callback when server sends ack message', function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        writeAckService.recieve({
                            topic: topic,
                            action: constants_1.RECORD_ACTION.WRITE_ACKNOWLEDGEMENT,
                            correlationId: correlationId,
                            originalAction: constants_1.RECORD_ACTION.CREATEANDUPDATE,
                            isWriteAck: true
                        });
                        return [4 /*yield*/, utils_1.PromiseDelay(1)];
                    case 1:
                        _a.sent();
                        sinon_1.assert.calledOnce(callbackSpy);
                        sinon_1.assert.calledWith(callbackSpy);
                        return [2 /*return*/];
                }
            });
        }); });
        it('doesn\'t call callback twice', function () { return __awaiter(void 0, void 0, void 0, function () {
            var msg;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        msg = {
                            topic: topic,
                            action: constants_1.RECORD_ACTION.WRITE_ACKNOWLEDGEMENT,
                            correlationId: correlationId,
                            originalAction: constants_1.RECORD_ACTION.CREATEANDUPDATE,
                            isWriteAck: true
                        };
                        writeAckService.recieve(msg);
                        writeAckService.recieve(msg);
                        return [4 /*yield*/, utils_1.PromiseDelay(1)];
                    case 1:
                        _a.sent();
                        sinon_1.assert.calledOnce(callbackSpy);
                        sinon_1.assert.calledWith(callbackSpy);
                        return [2 /*return*/];
                }
            });
        }); });
        it('calls ack callback with error when server sends error message', function () { return __awaiter(void 0, void 0, void 0, function () {
            var errorAction;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        errorAction = constants_1.RECORD_ACTION.MESSAGE_DENIED;
                        writeAckService.recieve({
                            topic: topic,
                            action: errorAction,
                            correlationId: correlationId,
                            originalAction: constants_1.RECORD_ACTION.CREATEANDUPDATE,
                            isAck: true,
                            isError: true
                        });
                        return [4 /*yield*/, utils_1.PromiseDelay(1)];
                    case 1:
                        _a.sent();
                        sinon_1.assert.calledOnce(callbackSpy);
                        sinon_1.assert.calledWith(callbackSpy, constants_1.RECORD_ACTION[errorAction]);
                        return [2 /*return*/];
                }
            });
        }); });
    });
});
