// @deepstream/server/dist/src/constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATES = void 0;
__exportStar(require("@deepstream/protobuf/dist/types/all"), exports);
__exportStar(require("@deepstream/protobuf/dist/types/messages"), exports);
var STATES;
(function (STATES) {
    STATES["CONFIG_LOADED"] = "CONFIG_LOADED";
    STATES["LOGGER_INIT"] = "LOGGER_INIT";
    STATES["SERVICE_INIT"] = "SERVICE_INIT";
    STATES["HANDLER_INIT"] = "HANDLER_INIT";
    STATES["CONNECTION_ENDPOINT_INIT"] = "CONNECTION_ENDPOINT_INIT";
    STATES["PLUGIN_INIT"] = "PLUGIN_INIT";
    STATES["RUNNING"] = "RUNNING";
    STATES["PLUGIN_SHUTDOWN"] = "PLUGIN_SHUTDOWN";
    STATES["CONNECTION_ENDPOINT_SHUTDOWN"] = "CONNECTION_ENDPOINT_SHUTDOWN";
    STATES["HANDLER_SHUTDOWN"] = "HANDLER_SHUTDOWN";
    STATES["SERVICE_SHUTDOWN"] = "SERVICE_SHUTDOWN";
    STATES["LOGGER_SHUTDOWN"] = "LOGGER_SHUTDOWN";
    STATES["STOPPED"] = "STOPPED";
})(STATES = exports.STATES || (exports.STATES = {}));
