// @deepstream/types/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var LOG_LEVEL;
(function (LOG_LEVEL) {
    LOG_LEVEL[LOG_LEVEL["DEBUG"] = 0] = "DEBUG";
    LOG_LEVEL[LOG_LEVEL["INFO"] = 1] = "INFO";
    LOG_LEVEL[LOG_LEVEL["WARN"] = 2] = "WARN";
    LOG_LEVEL[LOG_LEVEL["ERROR"] = 3] = "ERROR";
    LOG_LEVEL[LOG_LEVEL["FATAL"] = 4] = "FATAL";
    LOG_LEVEL[LOG_LEVEL["OFF"] = 100] = "OFF";
})(LOG_LEVEL = exports.LOG_LEVEL || (exports.LOG_LEVEL = {}));
class Handler {
    async close() { }
}
exports.Handler = Handler;
// tslint:disable-next-line: max-classes-per-file
class DeepstreamPlugin {
    async whenReady() { }
    async close() { }
}
exports.DeepstreamPlugin = DeepstreamPlugin;
var EVENT;
(function (EVENT) {
    EVENT["INFO"] = "INFO";
    EVENT["ERROR"] = "ERROR";
    EVENT["DEPRECATED"] = "DEPRECATED";
    EVENT["DEEPSTREAM_STATE_CHANGED"] = "DEEPSTREAM_STATE_CHANGED";
    EVENT["INCOMING_CONNECTION"] = "INCOMING_CONNECTION";
    EVENT["CLOSED_SOCKET_INTERACTION"] = "CLOSED_SOCKET_INTERACTION";
    EVENT["CLIENT_DISCONNECTED"] = "CLIENT_DISCONNECTED";
    EVENT["CONNECTION_ERROR"] = "CONNECTION_ERROR";
    EVENT["AUTH_ERROR"] = "AUTH_ERROR";
    EVENT["AUTH_RETRY_ATTEMPTS_EXCEEDED"] = "AUTH_RETRY_ATTEMPTS_EXCEEDED";
    EVENT["REGISTERING_USER"] = "REGISTERING_USER";
    EVENT["FATAL_EXCEPTION"] = "FATAL_EXCEPTION";
    EVENT["NOT_VALID_UUID"] = "NOT_VALID_UUID";
    EVENT["CONFIG_TRANSFORM"] = "CONFIG_TRANSFORM";
    EVENT["CONFIG_ERROR"] = "CONFIG_ERROR";
    EVENT["PLUGIN_ERROR"] = "PLUGIN_ERROR";
    EVENT["PLUGIN_INITIALIZATION_ERROR"] = "PLUGIN_INITIALIZATION_ERROR";
    EVENT["PLUGIN_INITIALIZATION_TIMEOUT"] = "PLUGIN_INITIALIZATION_TIMEOUT";
    EVENT["HTTP_REQUEST_TIMEOUT"] = "HTTP_REQUEST_TIMEOUT";
    EVENT["LOCK_RELEASE_TIMEOUT"] = "LOCK_RELEASE_TIMEOUT";
    EVENT["LOCK_REQUEST_TIMEOUT"] = "LOCK_REQUEST_TIMEOUT";
    EVENT["LEADING_LISTEN"] = "LEADING_LISTEN";
    EVENT["LOCAL_LISTEN"] = "LOCAL_LISTEN";
    EVENT["INVALID_CONFIG_DATA"] = "INVALID_CONFIG_DATA";
    EVENT["INVALID_STATE_TRANSITION"] = "INVALID_STATE_TRANSITION";
    EVENT["INVALID_LEADER_REQUEST"] = "INVALID_LEADER_REQUEST";
    EVENT["CLUSTER_LEAVE"] = "CLUSTER_LEAVE";
    EVENT["CLUSTER_JOIN"] = "CLUSTER_JOIN";
    EVENT["CLUSTER_SIZE"] = "CLUSTER_SIZE";
    EVENT["UNSUPPORTED_ACTION"] = "UNSUPPORTED_ACTION";
    EVENT["UNKNOWN_ACTION"] = "UNKNOWN_ACTION";
    EVENT["CLOSED_SOCKET"] = "CLOSED_SOCKET";
    EVENT["TELEMETRY_DEBUG"] = "TELEMETRY_DEBUG";
    EVENT["TELEMETRY_UNREACHABLE"] = "TELEMETRY_UNREACHABLE";
})(EVENT = exports.EVENT || (exports.EVENT = {}));
